File: 79-word-search/word-search.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        # looks like dfs
        # can you use DP? well technical you can. but it can be tricky because it's 4 directions
        # you can make subproblem with word, but maybe not with board

        # 1. first starting poitn
        # 2. dig
        # Time complexity is O(2m*n) one for finding starting point and, second to dig

        self.board = board
        self.word = word
        self.ROW = len(self.board)
        self.COL = len(self.board[0])
        for i in range(self.ROW):
            for j in range(self.COL):
                if self.board[i][j] == self.word[0]:
                    visited = set()
                    result = self.dig(i, j, self.word, visited)
                    if result:
                        return True
        return False

    def dig(self, i, j, word, visited):
        if word == "":
            return True
        if i < 0 or i == self.ROW or j < 0 or j == self.COL:
            return False
        if (i, j) in visited:
            return False
        
        if self.board[i][j] == word[0]:
            visited.add((i, j))
            result = self.dig(i-1, j, word[1:], visited.copy())
            result = self.dig(i+1, j, word[1:], visited.copy()) or result
            result = self.dig(i, j-1, word[1:], visited.copy()) or result
            result = self.dig(i, j+1, word[1:], visited.copy()) or result
            return result
        else:
            return False

File: 79-word-search/word-search.java
class Solution {
    private char[][] board;
    private int ROW;
    private int COL;
    
    public boolean exist(char[][] board, String word) {
        this.board = board;
        ROW = board.length;
        COL = board[0].length;
        boolean[][] path = new boolean[ROW][COL];
        for (int i=0; i < ROW; i++) {
            for (int j=0; j < COL; j++) {
                if (dfs(i, j, word, path)) {
                    return true;
                }
            }
        }
        return false;
    }

    public boolean dfs(int i, int j, String word, boolean[][] path) {
        if (word.isEmpty()) {
            return true;
        }
        if (i < 0 || i == ROW || j < 0 || j == COL) {
            return false;
        }
        if (path[i][j]) {
            return false;
        }
        char desiredC = word.charAt(0);
        char c = board[i][j];
        if (c == desiredC) {
            String newWord = word.substring(1);
            path[i][j] = true;
            int[][] directions = {{0, -1}, {0, 1}, {-1, 0}, {1, 0}};
            for (int[] direction: directions) {
                int newI = i + direction[0];
                int newJ = j + direction[1];
                if (dfs(newI, newJ, newWord, path)) {
                    return true;
                }
            }
            path[i][j] = false;            
        }
        return false;
    }
}